src/implementations/junoswap/pool.rs

```diff
@@ -0,0 +1,337 @@
+                input_token,
+                input_amount: offer_asset.amount,
+                min_output: minimum_out_amount,
+                expiration: None,
+            })?,
+        });
+
+        let event = Event::new("apollo/cw-dex/swap")
+            .add_attribute("type", "junoswap")
+            .add_attribute("offer_asset", format!("{:?}", offer_asset))
+            .add_attribute("ask_asset_info", format!("{:?}", ask_asset_info))
+            .add_attribute("minimum_out_amount", minimum_out_amount.to_string())
+            .add_attribute("recipient", recipient.to_string());
+
+        Ok(Response::new().add_message(swap).add_event(event))
+    }
+
+    fn get_pool_liquidity(&self, deps: Deps) -> Result<AssetList, CwDexError> {
+        let pool_info = self.query_info(&deps.querier)?;
+
+        Ok(JunoAssetList(vec![
+            JunoAsset {
+                info: pool_info.token1_denom.into(),
+                amount: pool_info.token1_reserve,
+            },
+            JunoAsset {
+                info: pool_info.token2_denom.into(),
+                amount: pool_info.token2_reserve,
+            },
+        ])
+        .into())
+    }
+
+    fn simulate_provide_liquidity(
+        &self,
+        deps: Deps,
+        asset: AssetList,
+    ) -> Result<Asset, CwDexError> {
+        let pool_info = self.query_info(&deps.querier)?;
+
+        // Get the assets in the pool from the assets sent in
+        let juno_assets: JunoAssetList = asset.try_into()?;
+        let token1 = juno_assets.find(pool_info.token1_denom.into())?;
+        let token2 = juno_assets.find(pool_info.token2_denom.into())?;
+
+        let expected_lps = juno_get_lp_token_amount_to_mint(
+            token1.amount,
+            pool_info.lp_token_supply,
+            pool_info.token1_reserve,
+        )?;
+
+        let token2_amount = juno_get_token2_amount_required(
+            token2.amount,
+            token1.amount,
+            pool_info.lp_token_supply,
+            pool_info.token2_reserve,
+            pool_info.token1_reserve,
+        )?;
+
+        if token2_amount > token2.amount {
+            return Err(CwDexError::Std(StdError::generic_err(
+                "Not enough token2 to provide liquidity",
+            )));
+        }
+
+        Ok(Asset {
+            info: AssetInfo::Cw20(deps.api.addr_validate(&pool_info.lp_token_address)?),
+            amount: expected_lps,
+        }
+        .into())
+    }
+
+    fn simulate_withdraw_liquidity(
+        &self,
+        deps: Deps,
+        asset: Asset,
+    ) -> Result<AssetList, CwDexError> {
+        let pool_info = self.query_info(&deps.querier)?;
+
+        // Calculate min tokens out
+        let share_ratio = Decimal::from_ratio(pool_info.lp_token_supply, asset.amount);
+        let min_token1 = (share_ratio * pool_info.token1_reserve).checked_sub(Uint128::one())?;
+        let min_token2 = (share_ratio * pool_info.token2_reserve).checked_sub(Uint128::one())?;
```

Comment:
This math does not match what is in junoswap `execute_remove_liquidity` function?

Reply:
fixed